sync/threadgroup_test.go

```diff
@@ -173,6 +166,40 @@ func TestThreadGroupRace(t *testing.T) {
 	}
 }
 
+func TestThreadGroupClosedOnStop(t *testing.T) {
+	var tg ThreadGroup
+	var closed bool
+	tg.OnStop(func() { closed = true })
+	if closed {
+		t.Fatal("close function should not have been called yet")
+	}
+	if err := tg.Stop(); err != nil {
+		t.Fatal(err)
+	}
+	if !closed {
+		t.Fatal("close function should have been called")
+	}
+
+	// Stop has already been called, so the close function should be called
+	// immediately
+	closed = false
+	tg.OnStop(func() { closed = true })
+	if !closed {
+		t.Fatal("close function should have been called immediately")
+	}
+
+	// this close function (which calls a goroutine) should also be called
+	// immediately since Stop has been called already
+	closedChan := make(chan struct{})
+	tg.OnStop(func() { go func() { closedChan <- struct{}{} }() })
```

Comment:
Channel receives unblock as soon as a channel is closed so you could write this line as
```go
tg.OnStop(func() { go close(closedChan) })
```
code golf